### PR TITLE
Add metric for expired TLS certificates

### LINKED ISSUE #4046: Add metric for expired TLS certificates
We could add a Micrometer gauge that becomes non-zero when the server's TLS certificates are expired, so that our users can monitor and fix it as soon as possible.

We could also add a gauge for 'time to expire' which becomes negative when the certificates are expired.

### PR DESCRIPTION
**Pull request by klurpicolo**
Motivation:

As described in #4046, we'd like to add gauges to indicate the server's TLS certificate's expiration.

Modifications:

- Create new class `ServerSslContextUtil` and move method `validateSslContext` and `buildSslContext` from `VirtualHostBuilder` to it so that we can use in `Server`
- For each certificate in certificate chain, will add the following gauge metric
- Add gauge metric `armeria.server.certificate.validity` which becomes 1 when certificate is expired, otherwise 0.
- Add gauge metric `armeria.server.certificate.validity.days` which provides duration before certificate is expired in days

Result:

- Closes #4046
- We can check certificate expiration using below metric 
```
# HELP armeria_server_certificate_validity 1 if certificate is in validity period, 0 if certificate is not in validity period
# TYPE armeria_server_certificate_validity gauge
# TYPE armeria_server_certificate_validity{common_name="localhost",} 1.0
# TYPE armeria_server_certificate_validity{common_name="test.root.armeria",} 0.0
```
- We can check time to expire using below metric
```
# HELP armeria_server_certificate_validity_days Duration in day before certificate expires which becomes -1 if certificate is expired
# TYPE armeria_server_certificate_validity_days gauge
armeria_server_certificate_validity_days{common_name="localhost",} 60.0
armeria_server_certificate_validity_days{common_name="test.root.armeria",} -1.0
```

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 1]
```diff
@@ -391,6 +412,44 @@ void setupVersionMetrics() {
...
+                Gauge.builder("armeria.server.certificate.expiration", certificateExpirationInfoProvider)
+                     .description("0 if certificate is not expired, 1 if certificate is expired")
+                     .tags("CN", commonName)
+                     .register(meterRegistry);
+
+                final Supplier<Number> timeToExpireProvider = () ->
+                        Duration.between(Instant.now(), x509Certificate.getNotAfter().toInstant()).getSeconds();
+                Gauge.builder("armeria.server.certificate.time.to.expire", timeToExpireProvider)
+                     .description("Duration in second before certificate expires which becomes negative " +
+                                  "if certificate is expired")
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:444]: - Precision
  Second precision seems too accurate for certification.
  Should we use days and return `-1` if the duration is negative?

- Meter naming
  JDK's `X509Certificate` uses `validity` as an official term. How about following the naming convention?
  https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/security/cert/X509Certificate.java#L153
  - `armeria.server.certificate.validity` could be 0 if certificate is expired.
  - `armeria.server.certificate.validity.days` to expose the remaining days?

File: core/src/main/java/com/linecorp/armeria/server/ServerSslContextUtil.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.util.ReferenceCountUtil;
+
+/**
+ * Utilities for server {@link SslContext}.
+ */
+public final class ServerSslContextUtil {
+    private ServerSslContextUtil() {}
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/ServerSslContextUtil.java:43]: Could you place the constructor below static methods?
https://armeria.dev/community/developer-guide#organize

File: core/src/main/java/com/linecorp/armeria/server/ServerSslContextUtil.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.util.ReferenceCountUtil;
+
+/**
+ * Utilities for server {@link SslContext}.
+ */
+public final class ServerSslContextUtil {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/ServerSslContextUtil.java:42]:
```suggestion
final class ServerSslContextUtil {
```

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 4]
```diff
@@ -131,6 +147,11 @@ public static ServerBuilder builder() {
 
         setupVersionMetrics();
 
+        if (config().sslContextMapping() != null) {
+            final SslContext sslContext = config().sslContextMapping().map(defaultHostname());
+            setupTlsMetrics(sslContext);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:152]: Should we set up metrics for all `SslContext`s? We can add the hostname as a tag.

File: core/src/test/resources/com/linecorp/armeria/server/expired-certificate_chain.pem [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+VTETMBEGA1UECAwKU29tZS1TdGF0ZTEhMB8GA1UECgwYSW50ZXJuZXQgV2lkZ2l0
+cyBQdHkgTHRkMRowGAYDVQQDDBF0ZXN0LnJvb3QuYXJtZXJpYTCBnzANBgkqhkiG
+9w0BAQEFAAOBjQAwgYkCgYEAs8aEWGHZN41zAj3iTZ3TtliFUMxJeYV+ezV+/gbK
+rLpGyxMv400ZWe3gzPGI1O6rcGLytihN4/QgS1IW4JJ4OgDxjragwC+VHYt5q5uB
+Mc+7nP0lEiop6TTWi3Zl+29fMC9/n4wr7GtJ0wHfbNv2oQ32ZjKDDQwpfXsFxzEd
+VFMCAwEAATANBgkqhkiG9w0BAQsFAAOBgQB0D7OjWpxkiNK1DiOPCQEhu9g0IpmA
+BnCqyBr/unLjq9QmQIaFS55QADeay3pzoIJ+W/eUGQ9mh2qVAwlXjAIuJ9ueY+UB
+XOWtq4E13D1AcdTmk2KdOArVFOShaD9j6FJi3ZxABspbeLO+kbWtETB3qyWAX4QM
+w1JoHvTFHgBE+Q==
+-----END TRUSTED CERTIFICATE-----
```
Comment (ikhoon) [inline core/src/test/resources/com/linecorp/armeria/server/expired-certificate_chain.pem:29]: Add EOF?

File: core/src/main/java/com/linecorp/armeria/server/Server.java (hunk 1 above)
Comment (klurpicolo) [reply core/src/main/java/com/linecorp/armeria/server/Server.java:444]: I agreed, will update as you suggested.

File: core/src/main/java/com/linecorp/armeria/server/ServerSslContextUtil.java (hunk 2 above)
Comment (klurpicolo) [reply core/src/main/java/com/linecorp/armeria/server/ServerSslContextUtil.java:43]: Noted

File: core/src/main/java/com/linecorp/armeria/server/Server.java (hunk 4 above)
Comment (klurpicolo) [reply core/src/main/java/com/linecorp/armeria/server/Server.java:152]: I'll iterate over `config.virtualHosts` and set metric per each, is that sounds good?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/Server.java:152]: It sounds good.

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 6]
```diff
@@ -391,6 +413,48 @@ void setupVersionMetrics() {
...
+
+                final Supplier<Number> certificateExpirationInfoProvider = () -> {
+                    try {
+                        x509Certificate.checkValidity();
+                    } catch (CertificateExpiredException | CertificateNotYetValidException e) {
+                        return 0;
+                    }
+                    return 1;
+                };
+                Gauge.builder("armeria.server.certificate.validity", certificateExpirationInfoProvider)
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:436]: How about using `ToDoubleFunction` that does not create additional `Number` objects.
```java
Gauge.builder("armeria.server.certificate.validity", x509Certificate, ceriticate -> { ... })
```

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 7]
```diff
@@ -391,6 +413,48 @@ void setupVersionMetrics() {
...
+                     .description("1 if certificate is in validity period, 0 if certificate is not in " +
+                                  "validity period")
+                     .tags("common_name", commonName, "hostname", hostname)
+                     .register(meterRegistry);
+
+                final Supplier<Number> timeToExpireProvider = () -> {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:442]: Ditto.

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 8]
```diff
@@ -391,6 +413,48 @@ void setupVersionMetrics() {
...
+                                                           x509Certificate.getNotAfter().toInstant());
+                    return diff.isNegative() ? -1 : diff.toDays();
+                };
+                Gauge.builder("armeria.server.certificate.validity.days", timeToExpireProvider)
+                     .description("Duration in day before certificate expires which becomes -1 " +
+                                  "if certificate is expired")
+                     .tags("common_name", commonName, "hostname", hostname)
+                     .register(meterRegistry);
+            } catch (Exception ex) {
+                logger.warn("Fail to setup certificate expiration, " + ex.getMessage());
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:453]:
```suggestion
                logger.warn("Failed to setup certificate expiration. hostname: {}", hostname, ex);
```

File: core/src/test/java/com/linecorp/armeria/server/ServerTest.java [hunk 9]
```diff
@@ -513,6 +523,142 @@ void versionMetrics() {
...
+
+        assertThatGauge(meterRegistry, CERT_VALIDITY_GAUGE_NAME, "localhost").isZero();
+        assertThatGauge(meterRegistry, CERT_VALIDITY_DAYS_GAUGE_NAME, "localhost").isEqualTo(-1);
+    }
+
+    @Test
+    void tlsMetricGivenCertificateChainExpired() {
+        final InputStream expiredCertificateChain = getClass()
+                .getResourceAsStream("expired-certificate_chain.pem");
+        final InputStream pk = getClass().getResourceAsStream("expire_pk.key");
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/server/ServerTest.java:647]: Should we replace `_` with `-`?
```suggestion
                .getResourceAsStream("expired-certificate-chain.pem");
        final InputStream pk = getClass().getResourceAsStream("expire-pk.key");
```

**Review by ikhoon — COMMENTED** [review]
Nice work!

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 10]
```diff
@@ -391,6 +412,46 @@ void setupVersionMetrics() {
...
+    /**
+     * Sets up gauge metric for each server certificate.
+     */
+    private void setupTlsMetrics(SslContext sslContext, String hostname) {
+        final MeterRegistry meterRegistry = config().meterRegistry();
+
+        final SSLSession sslSession = validateSslContext(sslContext);
+        for (Certificate certificate : sslSession.getLocalCertificates()) {
+            try {
+                final X509Certificate x509Certificate = (X509Certificate) certificate;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:424]: Could we `continue` if `certificate` is not `instanceof X509Certificate`?

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 11]
```diff
@@ -391,6 +412,46 @@ void setupVersionMetrics() {
...
+                         try {
+                             x509Cert.checkValidity();
+                         } catch (CertificateExpiredException | CertificateNotYetValidException e) {
+                             return 0;
+                         }
+                         return 1;
+                     })
+                     .description("1 if certificate is in validity period, 0 if certificate is not in " +
+                                  "validity period")
+                     .tags("common_name", commonName, "hostname", hostname)
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:437]:
```suggestion
                     .tags("common.name", commonName, "hostname", hostname)
```
(Micrometer convention)

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 12]
```diff
@@ -391,6 +412,46 @@ void setupVersionMetrics() {
...
+                     .register(meterRegistry);
+
+                Gauge.builder("armeria.server.certificate.validity.days", x509Certificate, x509Cert -> {
+                         final Duration diff = Duration.between(Instant.now(),
+                                                                x509Cert.getNotAfter().toInstant());
+                         return diff.isNegative() ? -1 : diff.toDays();
+                     })
+                     .description("Duration in day before certificate expires which becomes -1 " +
+                                  "if certificate is expired")
+                     .tags("common_name", commonName, "hostname", hostname)
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:447]:
```suggestion
                     .tags("common.name", commonName, "hostname", hostname)
```
(Micrometer convention)

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 13]
```diff
@@ -391,6 +412,46 @@ void setupVersionMetrics() {
...
+                     .register(meterRegistry);
+            } catch (Exception ex) {
+                logger.warn("Failed to setup certificate expiration. hostname: {}", hostname, ex);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:450]:
```suggestion
                logger.warn("Failed to set up TLS certificate metrics for a host: {}", hostname, ex);
```

File: core/src/main/java/com/linecorp/armeria/server/Server.java [hunk 14]
```diff
@@ -844,4 +905,10 @@ private static boolean isLocalPort(ServerPort serverPort) {
         final InetAddress address = serverPort.localAddress().getAddress();
         return address.isAnyLocalAddress() || address.isLoopbackAddress();
     }
+
+    private static String getCommonName(X509Certificate certificate) throws CertificateEncodingException {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/Server.java:909]: Could move this method right after `setupTlsMetrics()` for readability.

File: core/src/test/java/com/linecorp/armeria/server/CertificationMetricTest.java [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+              .build();
+
+        assertThatGauge(meterRegistry, CERT_VALIDITY_GAUGE_NAME, "localhost").isZero();
+        assertThatGauge(meterRegistry, CERT_VALIDITY_DAYS_GAUGE_NAME, "localhost").isEqualTo(-1);
+        assertThatGauge(meterRegistry, CERT_VALIDITY_GAUGE_NAME, "test.root.armeria").isOne();
+        assertThatGauge(meterRegistry, CERT_VALIDITY_DAYS_GAUGE_NAME, "test.root.armeria").isPositive();
+    }
+
+    private static AbstractDoubleAssert<?> assertThatGauge(MeterRegistry meterRegistry, String gaugeName,
+                                                           String cn) {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/CertificationMetricTest.java:178]: :+1:

**Review by trustin — COMMENTED** [review]
Excellent work, @klurpicolo! Left some minor comments.

**Review by ikhoon — APPROVED** [review]
Thanks, @klurpicolo! 🎉🎉🎉
Please address @trustin's comment. 🙏

File: core/src/main/java/com/linecorp/armeria/server/Server.java (hunk 10 above)
Comment (klurpicolo) [reply core/src/main/java/com/linecorp/armeria/server/Server.java:424]: I'm not sure if I understand your question correctly😅. Is what you mean is that we should skip setting up metric if it's not instance of `X509Certificate` as below, right?

```java
if (!(certificate instanceof X509Certificate)) {
    continue;
}
final X509Certificate x509Certificate = (X509Certificate) certificate;
final String commonName = getCommonName(x509Certificate);
```

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/Server.java:424]: > What do you mean is that we should skip setting up metric.

+1 It is.

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/Server.java:424]: Correct :+1:

**Review by trustin — APPROVED** [review]
Excellent job, @klurpicolo! :+1:🙇

File: core/src/main/java/com/linecorp/armeria/server/Server.java (hunk 10 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/Server.java:428]: Pushed a commit that adds the instanceof check to your branch. 🙂

**Review by jrhee17 — APPROVED** [review]
Looks like a very useful addition for armeria server users! Thank you @klurpicolo ! 🙇 👍 🙇

File: core/src/main/java/com/linecorp/armeria/server/Server.java (hunk 10 above)
Comment (klurpicolo) [reply core/src/main/java/com/linecorp/armeria/server/Server.java:425]: Thank you, @trustin. 🙏

**Pull request merged**